src/MarginTrading.Backend.Services/Services/OvernightSwapService.cs

```diff
@@ -0,0 +1,216 @@
+using MarginTrading.Backend.Services.Assets;
+using MarginTrading.Backend.Services.TradingConditions;
+using Newtonsoft.Json;
+
+namespace MarginTrading.Backend.Services
+{
+	/// <summary>
+	/// Take care of overnight swap calculation and charging.
+	/// </summary>
+	public class OvernightSwapService : IOvernightSwapService
+	{
+		private readonly IOvernightSwapCache _overnightSwapCache;
+		private readonly IAccountAssetsCacheService _accountAssetsCacheService;
+		private readonly IAccountsCacheService _accountsCacheService;
+		private readonly ICommissionService _commissionService;
+		private readonly IOvernightSwapStateRepository _overnightSwapStateRepository;
+		private readonly IOvernightSwapHistoryRepository _overnightSwapHistoryRepository;
+		private readonly IOrderReader _orderReader;
+		private readonly IThreadSwitcher _threadSwitcher;
+		private readonly AccountManager _accountManager;
+		private readonly MarginSettings _marginSettings;
+		private readonly ILog _log;
+		
+		private static readonly object LockObj = new object();
+
+		private DateTime _currentStartTimestamp;
+
+		public OvernightSwapService(
+			IOvernightSwapCache overnightSwapCache,
+			IAccountAssetsCacheService accountAssetsCacheService,
+			IAccountsCacheService accountsCacheService,
+			ICommissionService commissionService,
+			IOvernightSwapStateRepository overnightSwapStateRepository,
+			IOvernightSwapHistoryRepository overnightSwapHistoryRepository,
+			IOrderReader orderReader,
+			IThreadSwitcher threadSwitcher,
+			AccountManager accountManager,
+			MarginSettings marginSettings,
+			ILog log)
+		{
+			_overnightSwapCache = overnightSwapCache;
+			_accountAssetsCacheService = accountAssetsCacheService;
+			_accountsCacheService = accountsCacheService;
+			_commissionService = commissionService;
+			_overnightSwapStateRepository = overnightSwapStateRepository;
+			_overnightSwapHistoryRepository = overnightSwapHistoryRepository;
+			_orderReader = orderReader;
+			_threadSwitcher = threadSwitcher;
+			_accountManager = accountManager;
+			_marginSettings = marginSettings;
+			_log = log;
+		}
+
+		public void Start()
+		{
+			//initialize cache from storage
+			var savedState = _overnightSwapStateRepository.GetAsync().GetAwaiter().GetResult().ToList();
+			_overnightSwapCache.Initialize(savedState.Select(OvernightSwapCalculation.Create));
+			
+			//start calculation
+			CalculateAndChargeSwaps();
+			
+			//TODO if server was down more that a day.. calc N days
+		}
+
+		public void CalculateAndChargeSwaps()
+		{
+			lock (LockObj)
+			{
+				_currentStartTimestamp = DateTime.UtcNow;
+
+				//read orders syncronously
+				var openOrders = _orderReader.GetActive();
+
+				//filter orders that are already calculated
+				var calculatedIds = _overnightSwapCache.GetAll()
+					.Where(x => x.Time > LastInvocationTime)
+					.SelectMany(x => x.OpenOrderIds);
+				var filteredOrders = openOrders.Where(x => !calculatedIds.Contains(x.Id));
```

Comment:
And what about already closed orders?
If last calculation failed and user closed order? We need to find them in history or, at least send critical error to slack.

Reply:
Detection of such cases added.